### PR TITLE
[Feature]: Disable/Enable Cut, Save, Copy, Paste Flag

### PR DESCRIPTION
**Pull request by marcomaddy1982**
## What's new in this PR?

from the Feature-Flags configuration file it is now possible to enable and disable the cut, save, copy and paste action in the conversation and the camera roll action in the profile

## Dependency

build-configuration: https://github.com/wireapp/wire-ios-build-configuration/pull/28
JIRA Ticket: https://wearezeta.atlassian.net/browse/ZIOS-13657
JIRA Ticket: https://wearezeta.atlassian.net/browse/ZIOS-13659

### CODE REVIEW

File: Wire-iOS/Sources/UserInterface/Components/Views/MarkdownTextView.swift [hunk 1]
```diff
@@ -59,6 +60,24 @@ final class MarkdownTextView: NextResponderTextView {
...
+        switch action {
+        case #selector(UIResponderStandardEditActions.paste(_:)) where !sessionManager.isCopyAndPasteEnable,
+             #selector(UIResponderStandardEditActions.cut(_:)) where !sessionManager.isCopyAndPasteEnable,
+             #selector(UIResponderStandardEditActions.copy(_:)) where !sessionManager.isCopyAndPasteEnable:
+             return false
+        default:
+            //return true : this is not correct
+            return super.canPerformAction(action, withSender: sender)
+        }
+    }
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/UserInterface/Components/Views/MarkdownTextView.swift]: I think you could clean this up by:

```swift
switch action {
case #selector(UIResponderStandardEditActions.paste(_:)), ..., ...:
  if SessionManager.shared?.isCopyAndPasteEnable == false else { return false }
  fallthrough
default:
  return super.canPerformAction(action, withSender: sender)
}


```

File: Wire-iOS/Sources/UserInterface/Components/Views/MarkdownTextView.swift [hunk 2]
```diff
@@ -59,6 +60,24 @@ final class MarkdownTextView: NextResponderTextView {
...
+        guard let sessionManager = SessionManager.shared else {
+            return super.canPerformAction(action, withSender: sender)
+        }
+        switch action {
+        case #selector(UIResponderStandardEditActions.paste(_:)) where sessionManager.isDisabledClipboard,
+             #selector(UIResponderStandardEditActions.cut(_:)) where sessionManager.isDisabledClipboard,
+             #selector(UIResponderStandardEditActions.copy(_:)) where sessionManager.isDisabledClipboard:
+             return false
+        default:
+            //return true : this is not correct
```
Comment (billypchan) [inline Wire-iOS/Sources/UserInterface/Components/Views/MarkdownTextView.swift:75]: comment left?

File: Wire-iOS.xcodeproj/project.xcworkspace/xcshareddata/WorkspaceSettings.xcsettings [hunk 3]
```diff
@@ -1,5 +1,8 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
-<dict/>
+<dict>
+	<key>PreviewsEnabled</key>
```
Comment (billypchan) [inline Wire-iOS.xcodeproj/project.xcworkspace/xcshareddata/WorkspaceSettings.xcsettings:5]: It this still needed?

File: Wire-iOS/Resources/Configuration/Feature-Flags.xcconfig [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+// Configuration settings file format documentation can be found at:
+// https://help.apple.com/xcode/#/dev745c5c974
+
+CLIPBOARD_ENABLED=1
+SAVE_MESSAGE_ENABLED=1
+PROFILE_CAMERA_ROLL_ENABLED=1
```
Comment (johnxnguyen) [inline Wire-iOS/Resources/Configuration/Feature-Flags.xcconfig:24]: I suggest to rename this file `SecurityFlags` to reveal the intent better. Also a short description about the intent of these flags and perhaps a hint where to override these values would be nice.

File: Wire-iOS/Sources/Helpers/Bundle+FeatureFlags.swift [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    static var saveMessageEnabled: Bool {
+        return Bundle.appMainBundle.infoForKey("SaveMessageEnabled") == "1"
+    }
+    
+    static var profileCemeraRollEnabled: Bool {
+        return Bundle.appMainBundle.infoForKey("ProfileCemeraRollEnabled") == "1"
+    }
+}
+
+public enum FeatureFlag {
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Helpers/Bundle+FeatureFlags.swift:21]: Likewise, I would rename this to `SecurityFlag`.

File: Wire-iOS/Sources/Helpers/Bundle+FeatureFlags.swift [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  case clipboard
+  case save
+  case profileCamera
+  
+  public var isEnabled: Bool {
+    switch self {
+      case .clipboard:
+        return Bundle.clipboardEnabled
```
Comment (billypchan) [inline Wire-iOS/Sources/Helpers/Bundle+FeatureFlags.swift:43]: We can clean up these reported getter like:
```
return Bundle.appMainBundle.infoForKey(bundleKey) == "1"
```

```
var bundleKey: String {
switch self {
case .clipboard: return "ClipboardEnabled"
}
```

File: Wire-iOS/Sources/Helpers/Bundle+FeatureFlags.swift [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    }
+    
+    static var profileCemeraRollEnabled: Bool {
+        return Bundle.appMainBundle.infoForKey("ProfileCemeraRollEnabled") == "1"
+    }
+}
+
+public enum FeatureFlag {
+  case clipboard
+  case save
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Helpers/Bundle+FeatureFlags.swift:23]: I suggest a more descriptive name such as `saveMessage`.

File: Wire-iOS/Resources/Configuration/Feature-Flags.xcconfig (hunk 4 above)
Comment (johnxnguyen) [reply Wire-iOS/Resources/Configuration/Feature-Flags.xcconfig:24]: Also, I understand that we want a flag for all camera roll usage, not just in the profile screen. In which case, I suggest to rename this to `CAMERA_ROLL_ENABLED`.

File: Wire-iOS/Sources/Helpers/Bundle+FeatureFlags.swift [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+        return Bundle.appMainBundle.infoForKey("ClipboardEnabled") == "1"
+    }
+    
+    static var saveMessageEnabled: Bool {
+        return Bundle.appMainBundle.infoForKey("SaveMessageEnabled") == "1"
+    }
+    
+    static var profileCemeraRollEnabled: Bool {
+        return Bundle.appMainBundle.infoForKey("ProfileCemeraRollEnabled") == "1"
+    }
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Helpers/Bundle+FeatureFlags.swift]: Typo: "Cemera" in the property name, the key and also in the info plist.

File: Wire-iOS/Sources/UserInterface/Settings/ProfileSelfPictureViewController.swift [hunk 9]
```diff
@@ -91,6 +91,7 @@ final class ProfileSelfPictureViewController: UIViewController {
 
     private func addCameraButton() {
         
+        cameraButton.isHidden = !FeatureFlag.profileCamera.isEnabled
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/UserInterface/Settings/ProfileSelfPictureViewController.swift:94]: I think we want to hide the `libraryButton`, not the `cameraButton`.

File: Wire-iOS/Resources/Configuration/Feature-Flags.xcconfig (hunk 4 above)
Comment (marcomaddy1982) [reply Wire-iOS/Resources/Configuration/Feature-Flags.xcconfig:24]: About the `PROFILE_CAMERA_ROLL_ENABLED` there is also the camera roll in the keyboard to disable in another ticket so probably I would keep separated the two.

**Comment by johnxnguyen** [comment]
I still see some references to the xcconfig file, was it properly deleted?

**Comment by marcomaddy1982** [comment]
> I still see some references to the xcconfig file, was it properly deleted?

It is the default one but it will be overwritten every time

**Comment by johnxnguyen** [comment]
The configuration files that are pulled in via `setup.sh` are placed in a directory call `Configuration`, not in `Resources`.

**Comment by marcomaddy1982** [comment]
> The configuration files that are pulled in via `setup.sh` are placed in a directory call `Configuration`, not in `Resources`.

Yes yes I know, that one is the default value there in `Configuration`.

**Pull request merged**